tsl/src/compression/compression_storage.c

```diff
@@ -0,0 +1,330 @@
+
+	create->relation = compress_rel;
+	tbladdress = DefineRelation(create, RELKIND_RELATION, owner, NULL, NULL);
+	CommandCounterIncrement();
+	compress_relid = tbladdress.objectId;
+	ts_copy_relation_acl(ht->main_table_relid, compress_relid, owner);
+	ts_catalog_restore_user(&sec_ctx);
+	ts_hypertable_create_compressed(compress_relid, compress_hypertable_id);
+
+	return compress_hypertable_id;
+}
+
+Oid
+compression_chunk_create(Chunk *src_chunk, Chunk *chunk, List *column_defs, Oid tablespace_oid)
+{
+	ObjectAddress tbladdress;
+	CatalogSecurityContext sec_ctx;
+	Datum toast_options;
+	static char *validnsps[] = HEAP_RELOPT_NAMESPACES;
+	CompressionSettings *settings = ts_compression_settings_get(src_chunk->hypertable_relid);
+
+	Oid owner = ts_rel_get_owner(chunk->hypertable_relid);
+
+	CreateStmt *create;
+	RangeVar *compress_rel;
+
+	create = makeNode(CreateStmt);
+	create->tableElts = column_defs;
+	create->inhRelations = NIL;
+	create->ofTypename = NULL;
+	create->constraints = NIL;
+	create->options = NULL;
+	create->oncommit = ONCOMMIT_NOOP;
+	create->tablespacename = get_tablespace_name(tablespace_oid);
+	create->if_not_exists = false;
+
+	/* Invalid tablespace_oid <=> NULL tablespace name */
+	Assert(!OidIsValid(tablespace_oid) == (create->tablespacename == NULL));
+
+	/* create the compression table */
+	/* NewRelationCreateToastTable calls CommandCounterIncrement */
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+	compress_rel = makeRangeVar(NameStr(chunk->fd.schema_name), NameStr(chunk->fd.table_name), -1);
+
+	create->relation = compress_rel;
+	tbladdress = DefineRelation(create, RELKIND_RELATION, owner, NULL, NULL);
+	CommandCounterIncrement();
+	chunk->table_id = tbladdress.objectId;
+	ts_copy_relation_acl(chunk->hypertable_relid, chunk->table_id, owner);
+	toast_options =
+		transformRelOptions((Datum) 0, create->options, "toast", validnsps, true, false);
+	(void) heap_reloptions(RELKIND_TOASTVALUE, toast_options, true);
+	NewRelationCreateToastTable(chunk->table_id, toast_options);
+	ts_catalog_restore_user(&sec_ctx);
+	modify_compressed_toast_table_storage(settings, column_defs, chunk->table_id);
+
+	set_statistics_on_compressed_table(chunk->table_id);
+	set_toast_tuple_target_on_compressed(chunk->table_id);
+
+	create_compressed_table_indexes(chunk, settings);
+	return chunk->table_id;
+}
+
+static void
+set_toast_tuple_target_on_compressed(Oid compressed_table_id)
+{
+	DefElem def_elem = {
+		.type = T_DefElem,
+		.defname = "toast_tuple_target",
+		.arg = (Node *) makeInteger(128),
+		.defaction = DEFELEM_SET,
+		.location = -1,
+	};
+	AlterTableCmd cmd = {
+		.type = T_AlterTableCmd,
+		.subtype = AT_SetRelOptions,
+		.def = (Node *) list_make1(&def_elem),
+	};
+	ts_alter_table_with_event_trigger(compressed_table_id, NULL, list_make1(&cmd), true);
+}
+
+static void
+set_statistics_on_compressed_table(Oid compressed_table_id)
```

Comment:
Might be also good to rename to avoid confusion of internal compression hypertable vs compressed chunk.

Reply:
hmm its still a table, maybe change it to relation?

Reply:
I mean, it's called on a chunks only, not on the internal compression hypertable, right?

Reply:
Changed.